Panels/IdProvider.js

```diff
@@ -0,0 +1,110 @@
+import hoc from '@enact/core/hoc';
+import React from 'react';
+
+let GlobalId = 0;
+
+const ID_KEY = '$$ID$$';
+
+/**
+ * Default config for {@link agate/Panels.IdProvider}
+ *
+ * @hocconfig
+ * @memberof agate/Panels.IdProvider
+ */
+const defaultConfig = {
+	/**
+	 * Prop to pass the identifier generation function
+	 *
+	 * @type {String}
+	 * @default generateId
+	 * @memberof agate/Panels.IdProvider.defaultConfig
+	 */
+	generateProp: 'generateId',
+
+	/**
+	 * Prop to pass the identifier
+	 *
+	 * @type {String}
+	 * @default id
+	 * @memberof agate/Panels.IdProvider.defaultConfig
+	 */
+	idProp: 'id',
+
+	/**
+	 * Optional prefix for the identifier
+	 *
+	 * @type {String}
+	 * @default 'c_'
+	 * @memberof agate/Panels.IdProvider.defaultConfig
+	 */
+	prefix: 'c_'
+};
+
+/**
+ * A higher-order component that generates globally-unique identifiers
+ *
+ * @class IdProvider
+ * @hoc
+ * @private
+ * @memberof agate/Panels
+ */
+const IdProvider = hoc(defaultConfig, (config, Wrapped) => {
```

Comment:
We talked in Slack about this and I agree this could be moved somewhere else and refactor agate and moonstone to use those, like `ui` or maybe `core`?

Reply:
Yeah, we'll need to ultimately do some module organization. It looks like for the time being, we'll bundle them in agate until we are able to introduce enact API changes. I've left these alone for now.